src/interrupt/mod.rs

```diff
@@ -0,0 +1,193 @@
+//! in-band messages to replace traditional out-of-band assertion of dedicated interrupt lines.
+//! While more complex to implement in a device, message signaled interrupts have some significant
+//! advantages over pin-based out-of-band interrupt signaling. Message signaled interrupts are
+//! supported in PCI bus since its version 2.2, and in later available PCI Express bus. Some non-PCI
+//! architectures also use message signaled interrupts.
+//!
+//! On the other hand, IRQ is a term commonly used by Operating Systems when dealing with hardware
+//! interrupts. The IRQ number managed by OSes is independent of the IRQ number managed by VMM.
+//! To make things easy, `Interrupt Source` is used instead of IRQ for pin-based interrupts and
+//! MSI interrupts.
+//!
+//! A device may support multiple types of interrupts, and each type of interrupt may support one
+//! or multiple interrupt sources. For example, a PCI device may support:
+//! * Legacy Irq: exactly one interrupt source.
+//! * PCI MSI Irq: 1,2,4,8,16,32 interrupt sources.
+//! * PCI MSIx Irq: 2^n(n=0-11) interrupt sources.
+//!
+//! A distinct Interrupt Source Identifier (ISID) will be assigned to each interrupt source.
+//! An ID allocator will be used to allocate and free Interrupt Source Identifiers for devices.
+//! To decouple the vm-device crate from the ID allocator, the vm-device crate doesn't take the
+//! responsibility to allocate/free Interrupt Source IDs but only makes use of assigned IDs.
+//!
+//! The overall flow to deal with interrupts is:
+//! * the VMM creates an interrupt manager
+//! * the VMM creates a device manager, passing on an reference to the interrupt manager
+//! * the device manager passes on an reference to the interrupt manager to all registered devices
+//! * guest kernel loads drivers for virtual devices
+//! * guest device driver determines the type and number of interrupts needed, and update the
+//!   device configuration
+//! * the virtual device backend requests the interrupt manager to create an interrupt group
+//!   according to guest configuration information
+
+use std::sync::Arc;
+use vmm_sys_util::eventfd::EventFd;
+
+/// Reuse std::io::Result to simplify inter operability.
+pub type Result<T> = std::io::Result<T>;
+
+/// Data type to store an interrupt index.
+pub type InterruptIndex = u32;
+
+#[cfg(feature = "generic_msi")]
+/// Maximum number of generic Message Signalled Interrupts supported per device.
+pub const MAX_GENERIC_MSI_IRQS: InterruptIndex = 32;
+
+/// Type of interrupt source.
+pub enum InterruptSourceType {
+    #[cfg(feature = "legacy_irq")]
+    /// Legacy Pin-based Interrupt.
+    /// On x86 platforms, legacy interrupts are routed through 82599 PICs and/or IOAPICs.
+    Legacy,
+    #[cfg(feature = "generic_msi")]
+    /// Generic Message Signalled Interrupt.
+    /// Some non-PCI devices (like HPET on x86) make use of generic MSI in platform specific ways.
```

Comment:
Is `generic MSI` a real thing? I mean I know about very specific HPET use case, but is it standardized in any way? As a started, we should stick to the 3 other options, and leave the very specific HPET use case aside.

Reply:
We are extending the Virtio MMIO transport layer to support MSI, so used "generic MSI" for it.
Or we may called it "MmioMsi", something similar to HPET, DMAR or platform MSI on arm.

Reply:
Yes `MmioMsi` sounds a bit better to me, as it contrasts with PCI. But the thing is that you're going to call something `MmioMsi` when it's not part of a spec yet. I'm worried that we're mixing very well defined types with other ones that are not yet approved by any upstream, WDYT?

Reply:
Sadly, that's the truth:(
How about first have some concrete implementation and benchmark data, then we could propose to update viritio spec.

Reply:
Sounds good!